### PR TITLE
feat(perf): integrate OnStateHook in executor

### LINKED ISSUE #11330: Integrate OnStateHook in executor
This should unblock #11161

with #11068 we can now hook into state updates on system calls, we also need this for all txs in the executor.

this can be achieved by also introducing the hook generic on the executor types, similar to what we did with:

https://github.com/paradigmxyz/reth/blob/3297dcb4eb0e26eeac7940ec7adcd64b8ee19fbc/crates/evm/src/execute.rs#L37-L45

but accepting a OnStateHook instance.

then we'd also need to call the hook before we commit the regular txs
https://github.com/paradigmxyz/reth/blob/3297dcb4eb0e26eeac7940ec7adcd64b8ee19fbc/crates/ethereum/evm/src/execute.rs#L189-L189

### PR DESCRIPTION
**Pull request by fgimenez**
Closes: #11330 

It can be used for instance with state root calculation like this
````rust
let (tx, rx) = channel();
// spawn SR
p.executor().execute_with_state_hook(.., |state| tx.send(state.clone()));
````

### CODE REVIEW

File: crates/ethereum/evm/src/execute.rs [hunk 1]
```diff
@@ -169,7 +172,8 @@ where
                     error: Box::new(new_err),
                 }
             })?;
-            evm.db_mut().commit(state);
+            evm.db_mut().commit(state.clone());
+            system_caller.on_state(&ResultAndState { result: result.clone(), state });
```
Comment (mattsse) [inline crates/ethereum/evm/src/execute.rs:176]: these should be swapped, and we can delay the resultandstate destruct on L167, then we can get rid of the clone

File: crates/ethereum/evm/src/execute.rs [hunk 2]
```diff
@@ -260,25 +264,37 @@ where
...
+    }
+
     /// Execute a single block and apply the state changes to the internal state.
     ///
     /// Returns the receipts of the transactions in the block, the total gas used and the list of
     /// EIP-7685 [requests](Request).
     ///
     /// Returns an error if execution fails.
-    fn execute_without_verification(
+    fn execute_without_verification_with_state_hook<F>(
```
Comment (mattsse) [inline crates/ethereum/evm/src/execute.rs:281]: ah cool, this is smarter than integrating this in the Executor type as field

File: crates/evm/src/system_calls/mod.rs [hunk 3]
```diff
@@ -321,4 +318,11 @@ where
 
         eip7251::post_commit(result_and_state.result)
     }
+
+    /// Delegate to stored `OnStateHook`, noop if hook is `None`.
+    pub fn on_state(&mut self, state: &ResultAndState) {
+        if let Some(ref mut hook) = &mut self.hook {
+            hook.on_state(state);
+        }
+    }
```
Comment (mattsse) [inline crates/evm/src/system_calls/mod.rs:327]: I think for now this is perfectly fine, I assume there's a better approach to this once we try to redesign the executor parts

**Review by mattsse — CHANGES_REQUESTED** [review]
cool, this will do it, 
I wonder if the hook generic will just be code bloat, so we could consider using dyn here

File: crates/ethereum/evm/src/execute.rs (hunk 1 above)
Comment (fgimenez) [reply crates/ethereum/evm/src/execute.rs]: good catch, done

**Review by mattsse — APPROVED** [review]

**Pull request merged**